### PR TITLE
Multi token features

### PR DESCRIPTION
**Pull request by harture**

### CODE REVIEW

File: api/account/api.go [hunk 1]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
 	return nil
 }
 
+// Validate is a validator for UpdatePasswordBody
+func (updatePwd UpdatePasswordBody) Validate() error {
+	if !matchesRegExp(updatePwd.CurrentPassword, RegExpPassword) {
+		return errors.New("Invalid current Password")
```
Comment (bsoniam) [inline api/account/api.go]: TODO: will trigger problems with the current error formatting

File: api/account/api.go [hunk 2]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+	}
+
+	if !matchesRegExp(updatePwd.NewPassword, RegExpPassword) {
+		return errors.New("Invalid new Password")
```
Comment (bsoniam) [inline api/account/api.go]: TODO

File: api/account/api.go [hunk 3]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+	}
+
+	if !matchesRegExp(updatePwd.ConfirmPassword, RegExpPassword) {
+		return errors.New("Invalid confirm Password")
```
Comment (bsoniam) [inline api/account/api.go]: TODO

File: api/account/api.go [hunk 4]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+	}
+
+	return nil
+}
+
+// Validate is a validator for CredentialRepresentation
+func (credential CredentialRepresentation) Validate() error {
+	if credential.Id != nil && !matchesRegExp(*credential.Id, RegExpID) {
+		return errors.New("Invalid Id")
```
Comment (bsoniam) [inline api/account/api.go]: TODO

File: api/account/api.go [hunk 5]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+	}
+
+	if credential.Type != nil && !matchesRegExp(*credential.Type, RegExpType) {
+		return errors.New("Invalid Type")
```
Comment (bsoniam) [inline api/account/api.go]: TODO

File: api/account/api.go [hunk 6]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+	}
+
+	if credential.UserLabel != nil && !matchesRegExp(*credential.UserLabel, RegExpLabel) {
+		return errors.New("Invalid Label")
```
Comment (bsoniam) [inline api/account/api.go]: TODO

File: pkg/account/component.go [hunk 7]
```diff
@@ -187,3 +198,105 @@ func (c *component) DeleteAccount(ctx context.Context) error {
...
+
+	if err != nil {
+		c.logger.Warn("err", err.Error())
+		return err
+	}
+
+	//store the API call into the DB
+	// the error should be treated
+	additionalInfos, _ := json.Marshal(map[string]string{"credentialID": credentialID, "label": label})
+	_ = c.reportEvent(ctx, "SELF_UPDATE_CREDENTIAL", database.CtEventRealmName, currentRealm, database.CtEventUserID, userID, database.CtEventUsername, username, database.CtEventAdditionalInfo, string(additionalInfos))
```
Comment (bsoniam) [inline pkg/account/component.go]: TODO: log the event if the reportEvent fails (as it is done on master)

File: pkg/account/component.go [hunk 8]
```diff
@@ -187,3 +198,105 @@ func (c *component) DeleteAccount(ctx context.Context) error {
...
+
+	if err != nil {
+		c.logger.Warn("err", err.Error())
+		return err
+	}
+
+	//store the API call into the DB
+	// the error should be treated
+	additionalInfos, _ := json.Marshal(map[string]string{"credentialID": credentialID})
+	_ = c.reportEvent(ctx, "SELF_DELETE_CREDENTIAL", database.CtEventRealmName, currentRealm, database.CtEventUserID, userID, database.CtEventUsername, username, database.CtEventAdditionalInfo, string(additionalInfos))
```
Comment (bsoniam) [inline pkg/account/component.go]: TODO: log the event if the reportEvent fails (as it is done on master)

File: pkg/account/component.go [hunk 9]
```diff
@@ -187,3 +198,105 @@ func (c *component) DeleteAccount(ctx context.Context) error {
...
+
+	if err != nil {
+		c.logger.Warn("err", err.Error())
+		return err
+	}
+
+	//store the API call into the DB
+	// the error should be treated
+	additionalInfos, _ := json.Marshal(map[string]string{"credentialID": credentialID, "previousCredentialID": previousCredentialID})
+	_ = c.reportEvent(ctx, "SELF_MOVE_CREDENTIAL", database.CtEventRealmName, currentRealm, database.CtEventUserID, userID, database.CtEventUsername, username, database.CtEventAdditionalInfo, string(additionalInfos))
```
Comment (bsoniam) [inline pkg/account/component.go]: TODO: log the event if the reportEvent fails (as it is done on master)

File: pkg/account/component_test.go [hunk 10]
```diff
@@ -23,10 +24,10 @@ func genericUpdatePasswordTest(t *testing.T, oldPasswd, newPasswd, confirmPasswo
 	var mockCtrl = gomock.NewController(t)
 	defer mockCtrl.Finish()
 
-	mockKeycloakClient := mock.NewAccKeycloakClient(mockCtrl)
+	mockKeycloakAccountClient := mock.NewKeycloakAccountClient(mockCtrl)
 	mockEventDBModule := mock.NewEventsDBModule(mockCtrl)
-	mockLogger := mock.NewLogger(mockCtrl)
-	component := NewComponent(mockKeycloakClient, mockEventDBModule, mockLogger)
+	mockLogger := log.NewNopLogger()
```
Comment (bsoniam) [inline pkg/account/component_test.go]: Why going back to NewNopLogger() ?

File: pkg/account/endpoint.go [hunk 11]
```diff
@@ -38,17 +48,77 @@ type AccountComponent interface {
...
-		var body UpdatePasswordBody
+		var body api.UpdatePasswordBody
 
 		err := json.Unmarshal([]byte(m["body"]), &body)
 		if err != nil {
 			return nil, errrorhandler.CreateBadRequestError(internal.MsgErrInvalidParam + "." + internal.Body)
 		}
 
+		if err = body.Validate(); err != nil {
+			return nil, errrorhandler.CreateBadRequestError(err.Error())
```
Comment (bsoniam) [inline pkg/account/endpoint.go]: TODO

File: pkg/account/endpoint.go [hunk 12]
```diff
@@ -38,17 +48,77 @@ type AccountComponent interface {
...
+// MakeUpdateLabelCredentialEndpoint make the UpdateLabelCredential endpoint to set a new label for a credential.
+func MakeUpdateLabelCredentialEndpoint(component AccountComponent) cs.Endpoint {
+	return func(ctx context.Context, req interface{}) (interface{}, error) {
+		var m = req.(map[string]string)
+
+		var body api.CredentialRepresentation
+
+		err := json.Unmarshal([]byte(m["body"]), &body)
+		if err != nil {
+			return nil, errrorhandler.CreateBadRequestError("Invalid body")
```
Comment (bsoniam) [inline pkg/account/endpoint.go]: TODO

File: pkg/account/endpoint.go [hunk 13]
```diff
@@ -38,17 +48,77 @@ type AccountComponent interface {
...
+		}
+
+		if err = body.Validate(); err != nil {
+			return nil, errrorhandler.CreateBadRequestError(err.Error())
```
Comment (bsoniam) [inline pkg/account/endpoint.go]: TODO

File: pkg/account/endpoint.go [hunk 14]
```diff
@@ -38,17 +48,77 @@ type AccountComponent interface {
...
+		}
+
+		if body.UserLabel == nil {
+			return nil, errrorhandler.CreateBadRequestError("User label missing")
```
Comment (bsoniam) [inline pkg/account/endpoint.go]: TODO

File: Gopkg.toml [hunk 15]
```diff
@@ -31,7 +31,8 @@
 
 [[constraint]]
   name = "github.com/cloudtrust/keycloak-client"
-  version = "v1.0-rc8"
+  #version = "v1.0-rc7"
+  branch="multi-token"
```
Comment (fperot74) [inline Gopkg.toml]: Remove before merging ? (do we keep it in a branch while production is still in 6.0.1 ?)

File: api/account/api.go [hunk 16]
```diff
@@ -85,13 +120,54 @@ func (user AccountRepresentation) Validate() error {
...
+
 func matchesRegExp(value, re string) bool {
 	res, _ := regexp.MatchString(re, value)
 	return res
 }
 
 // Regular expressions for parameters validation
 const (
+	RegExpID         = `^[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$`
+	RegExpIDNullable = `^([a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12})|(null)$`
```
Comment (fperot74) [inline api/account/api.go]: We want to support the explicit value (null) for null values ?

File: api/management/api.go [hunk 17]
```diff
@@ -98,19 +102,12 @@ func ConvertCredential(credKc *kc.CredentialRepresentation) CredentialRepresenta
...
-			value, ok := (*credKc.Config)[key]
-			if ok {
-				m[key] = value
-			}
-		}
-		cred.Config = &m
-	}
+	cred.CredentialData = credKc.CredentialData
+	cred.Temporary = credKc.Temporary
+	cred.Value = credKc.Value
```
Comment (fperot74) [inline api/management/api.go]: Do we want to transmit the Value ?

Comment (harture) [reply api/management/api.go]: Value is used to transmit raw password during password change

File: api/account/api.go (hunk 16 above)
Comment (harture) [reply api/account/api.go]: we use it only for credential move. If we move a credential at the top of the list, we call moveAfter with null as there is nothing before.

File: Gopkg.toml (hunk 15 above)
Comment (harture) [reply Gopkg.toml]: We will merge in master.

**Pull request merged**